changefeedccl: add non-batching retries to kafka sink

samiskin opened the pull request:
Resolves https://github.com/cockroachdb/cockroach/issues/80313                                  
                                    
A recurring painpoint of changefeeds is the "message too large" kafka
issue where a fixed size batch may sometimes end up being too many bytes
in size due to varying message sizes, but there was no mechanism to
reduce the batch size dynamically, so the changefeed would simply fail.
                                                                         
This PR adds an internal retry mechanism to the kafka sink, where upon
observing a messagetoolarge error the sink will pause all new incoming
messages, attempt to retry those messages with a progressively smaller
batch size, then once one is found resumes operation with the original
configuration.
                                                                         
Release note (bug fix): Changefeeds emitting to Kafka upon receiving a
"message too large" error will now halve the size of their batches until
it either succeeds or a batch size of 1 fails.

Release justification: high priority bug fix, new code is mostly only ran in 
cases where an unrecoverable error would've occurred previously and 
can be toggled with a cluster setting.

samiskin commented:
The summary of the solution here is:

1. Observe the first messagetoolarge error
2. block all new emits to the producer
3. Store all in-flight erroneous messages in a buffer until no more are in flight
4. Swap the kafka client and producer to one that does not batch (no messages are in-flight now so it should be safe)
5. Re-emit all buffered messages and wait for none to be in-flight again
6. Swap the kafka client and producer to the original config (again, nothing in flight)
7. Allow new emits once again.

samiskin commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -7089,3 +7089,67 @@ func TestSchemachangeDoesNotBreakSinklessFeed(t *testing.T) {
+	defer utilccl.TestingEnableEnterprise()()
+
+	skip.UnderRaceWithIssue(t, 84736)
+	skip.UnderStress(t)
```
Looking into this rn ⌛

miretskiy commented on pkg/ccl/changefeedccl/sink_kafka.go:
```diff
@@ -167,39 +197,83 @@ func defaultSaramaConfig() *saramaConfig {
+
+	// Start the worker
+	if s.clientMu.stopWorkerCh != nil {
+		close(s.clientMu.stopWorkerCh)
```
Should this be done in the maybeCloseClient?

HonoreDB reviewed (changes requested):
> Store all in-flight erroneous messages in a buffer until no more are in flight
This might be okay but makes me nervous, are you sure it's necessary? (vs. starting to retry as soon as we enter this state).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-:-N9XNOnu-QXVgEisAZZP:bsolqti)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo and @samiskin)

___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XIYNP3AFKaYp9zAAx:-N9XIYNP3AFKaYp9zAAy:b3xih99) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/changefeed_test.go#L7146)):*
> ```Go
> 				}
> 
> 				if !strings.Contains(status, "kafka server: Message was too large, server rejected it to avoid allocation error.") {
> ```

nonblocking: this makes me wish we had either a version of assert/require that returned an error instead of taking a testing.T, or a version of SucceedsSoon that passed in a special Fataler so that you could just use require here.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 435 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XLf9HA9ZkGl1kZ5Mn:-N9XLf9HA9ZkGl1kZ5Mo:b-ink9ob) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L435)):*
> ```Go
> 	}
> 
> 	s.clientMu.Lock()
> ```

Curious why this Lock is necessary (probably something I don't understand about go).
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 453 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9X8-ek7EXqHbiVjzHA:-N9X8-ek7EXqHbiVjzHB:bor5m2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L453)):*
> ```Go
> func isInternalRetryable(err error) bool {
> 	var kError sarama.KError
> 	return err != nil && errors.As(err, &kError) && kError == sarama.ErrMessageSizeTooLarge
> ```

Probably better to rely on error not being nil being checked at the callsite; I don't think we want to end up calling this method (and allocating a sarama.KError) on every message.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 499 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XJP5MDnB9RNBUMG7s:-N9XJP5MDnB9RNBUMG7t:bi7o44j) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L499)):*
> ```Go
> 		// If the message is to be buffered to resend just add it to the buffer
> 		// rather than clean it up or surface its error.
> 		if s.mu.isBufferingRetries && ackError != nil {
> ```

Unless you're doing this elsewhere I think we should check that ackMsg is a real message (has a non-nil Key and Value). We can also get errors for sentinel messages on this channel and those should not be retried.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 515 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XK8we-OxDKMfe7cFc:-N9XK8we-OxDKMfe7cFd:b-fhvfuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L515)):*
> ```Go
> 		}
> 
> 		// If no more messages are are inflight while buffering for retry, signal to
> ```

typo: are are
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XMvZAARAqvUXz1H7g:-N9XMvZAARAqvUXz1H7h:b-ehb08m) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L557)):*
> ```Go
> 		// Switch to sending messages one by one
> 		cfg := disableBatching(s.kafkaCfg)
> 		if s.resetClient(cfg) != nil {
> ```

Does this end up swallowing the error or does it get surfaced somewhere?


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 420 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9X6I_V83bTsUgE6oqC:-N9X6I_V83bTsUgE6oqD:bann24y) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L420)):*
> ```Go
> 	s.mu.Unlock()
> 	if retryCh != nil {
> 		log.Infof(ctx, "waiting for retry (%d messages inflight)", inflight)
> ```

consider adding
```
defer func(start time.Time) {
   log.Infof(ctx, "waited for %s", timeutil.Since(start))
}(timeutil.Now())
```
(Note: I always worry about use of ctx in defer -- you might need to use contet.Background() because ctx.Done might be true)...
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 437 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XFTYACmmpHvsnTd7e:-N9XFTYACmmpHvsnTd7f:b-k6cz6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L437)):*
> ```Go
> 	s.clientMu.Lock()
> 	producer := s.clientMu.producer
> 	s.clientMu.Unlock()
> ```

I'll keep reading, but this looks worrisome to me... Like what happens if the producer is swapped
and now the code below will block forever.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 492 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9X4mVe33pq6GUWXrok:-N9X4mVe33pq6GUWXrol:b2e0nuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L492)):*
> ```Go
> 		// Note that errors seen for the re-emitted messages are not retried a
> 		// second time and are surfaced just like any other sink error.
> 		if s.mu.retrySignalCh == nil && isInternalRetryable(ackError) {
> ```

This needs to be louder.  log.Error; plus add metrics.

Also, if the current batch size is 1 -- we definitely need to spit out permanent error.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 588 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XGY9H8PYPNZR4Z1xx:-N9XGY9H8PYPNZR4Z1xy:b-4z6wve) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L588)):*
> ```Go
> 			return
> 		case <-retryFlushCh:
> 			log.Infof(s.ctx, "retryWorker done waiting for retry flush")
> ```

could emit how much time this retry took.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
Had offline discussion re current implementation.  Look forward to seeing a more streamline solution.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-:-N9XseBO9iLxljdzGX1n:be551x2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo and @samiskin)



<!-- Sent from Reviewable.io -->

shermanCRL commented:
A note to think about some adversarial testing here, let’s exercise this thing. What if this situation is hit repeatedly, and we’ve adjusted repeatedly, and under high load.

samiskin committed:
changefeedccl: add non-batching retries to kafka sink

A recurring painpoint of changefeeds is the "message too large" kafka
issue where a fixed size batch may sometimes end up being too many bytes
in size due to varying message sizes, but there was no mechanism to
reduce the batch size dynamically, so the changefeed would simply fail.

This PR adds an internal retry mechanism to the kafka sink, where upon
observing a messagetoolarge error the sink will pause all new incoming
messages, attempt to retry those messages with a progressively smaller
batch size, then once one is found resumes operation with the original
configuration.

Release note (bug fix): Changefeeds emitting to Kafka upon receiving a
"message too large" error will now halve the size of their batches until
it either succeeds or a batch size of 1 fails.

Release justification: high priority bug fix, new code is mostly only ran in
cases where an unrecoverable error would've occurred previously and can
be toggled with a cluster setting.

samiskin commented:
More logging and comments and likely testing are needed but @miretskiy I just pushed the updated code for a first review pass on it

miretskiy reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo and @samiskin)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cNY6KEAr0IrQfMI-K):*
I think pr description needs to be updated since retries are now batching.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cO-iH-kgI4a5K1Eo7:-N9cO-iH-kgI4a5K1Eo8:buhpkxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L107)):*
> ```Go
> 	stopWorkerCh chan struct{}
> 
> 	clientMu struct {
> ```

let's add a comment why we need this mutex.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 451 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cahyh-QxPHHlLq0JV:-N9cahyh-QxPHHlLq0JW:b-uj0i2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L451)):*
> ```Go
> 	// retry internally with a more finely grained client until one is found that
> 	// can successfully send the errored messages.
> 	isBufferingRetries := false
> ```

you might not need this variable if you initialize retryBuf in startInternalRetry; then check for retyrBuf!=nil
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 457 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cZf90586ptXeIowU9:-N9cZf90586ptXeIowUA:bflda1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L457)):*
> ```Go
> 		retryErr = err
> 		isBufferingRetries = true
> 		s.clientMu.Lock() // Lock the client so that no new emits can occur
> ```

as mentioned below, I'm concerned that early termination of this function may
result in lock being held, and caller unable to even notice the error (because it needs to acquire lock).
I think the following is safer:
```
type maybeLocker struct {
  wrapped sync.Locker
  locked bool
}
func (l *maybeLocker) Lock() {
   l.wrapped.Lock()
   l.locked = true
}
func (l *maybeLocker) Unlock() {
   if l.locked {
     l.wrapped.Unlock()
     l.locked = false
   }
}

...
clientMu := &maybeLocker{s.clientMu}
defer clientMu.Unlock()

startInternalRetry := func() { 
    clientMu.Lock()
}

```

Or.. if you don't like maybeLock helper, you can do that with functions...
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 459 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cQLQA7Er9Pq2zd_-M:-N9cQLQA7Er9Pq2zd_-N:biaad0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L459)):*
> ```Go
> 		s.clientMu.Lock() // Lock the client so that no new emits can occur
> 	}
> 	retryBufferedMessages := func() {
> ```

lets 
```
s.clientMu.AssertHeld
```
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 464 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cZ1ND9D6hr42LFbBQ:-N9cZ1ND9D6hr42LFbBR:b2e1so1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L464)):*
> ```Go
> 		err := s.clientMu.client.Close()
> 		if err != nil {
> 			s.clientMu.clientErr = err
> ```

should we propagate error here? We're "throwing" the client away  and going to replace...
I wonder if this should be silent; or perhaps log.error
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 477 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cQU1Y4iTLpDeQwSNX:-N9cQU1Y4iTLpDeQwSNY:b-i6asm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L477)):*
> ```Go
> 		retryBuf = nil
> 
> 		s.kafkaCfg = validClient.Config()
> ```

is this the only place .Config() is used? If so, consider just returning config form handleBuffeedRetries?
you could event pass current config (ptr) to handleBufferedRetries so that it mutates config directly,
w/out the need to create new config object.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 480 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cZBBLDDoVpXgCKJS4:-N9cZBBLDDoVpXgCKJS5:bpj63bn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L480)):*
> ```Go
> 		s.clientMu.client = validClient
> 		s.clientMu.producer = validProducer
> 		producer = s.clientMu.producer
> ```

I would prefer if this function (retryBufferedMessages) was removed, and 
this logic was done at the call site:
```
  if isBufferingRetries && s.mu.inflight == 0 {
     		validClient, validProducer, err := s.handleBufferedRetries(retryBuf, retryErr)
		if err != nil {
			s.clientMu.clientErr = err
			return
		}
                if err := s.updateClient(validClient, validProducer); err != nil {
                     ....
                }
  }
```
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cR1TaDqNUmhBbb-pu:-N9cR1TaDqNUmhBbb-pv:b-jpdisr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L510)):*
> ```Go
> 
> 		if !isBufferingRetries && isInternalRetryable(ackError) {
> 			startInternalRetry(ackError)
> ```

I'm concerned that startInternalRetry grabs a lock, and until inflight == 0, we don't release the lock.
Yet, it's possible that some other error may occur, causing early termination of this go routine...
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 555 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9c_4QWETtaed2aFJe8:-N9c_4QWETtaed2aFJe9:b7utcuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L555)):*
> ```Go
> 	s.clientMu.AssertHeld()
> 
> 	lastSendErr := retryErr
> ```

I don't know why we need to pass in retryErr in the first place...
Do we need it?
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 564 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cUWqR8iWWa4hOZisp:-N9cUWqSCb29exiUZuId:b-h6els3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L564)):*
> ```Go
> 		// Surface the error if its not retryable or we weren't able to reduce the
> 		// batching config any further
> 		if !isInternalRetryable(lastSendErr) || newConfig.Producer.Flush == activeConfig.Producer.Flush {
> ```

I'd rather have reduceBatchingConfig return a bool indicating that it couldn't be reduced.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 585 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cUkrC-ZW7dvRpkR3C:-N9cUkrC-ZW7dvRpkR3D:bqyfp9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L585)):*
> ```Go
> 			break
> 		} else {
> 			_ = newProducer.Close()
> ```

definitely make it loud.  log error at least; 
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 606 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cU53lEJFmZFU9Wrax:-N9cU53lEJFmZFU9Wray:b-k6ugwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L606)):*
> ```Go
> func reduceBatchingConfig(c *sarama.Config) *sarama.Config {
> 	flooredHalve := func(num int) int {
> 		if num < 2 { // don't change if 1 or 0
> ```

does `return (num+1)>>1` work?


<!-- Sent from Reviewable.io -->

samiskin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 555 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9c_4QWETtaed2aFJe8:-N9gZEn39RZD1q_rr6NX:b-8x9ft5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L555)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't know why we need to pass in retryErr in the first place...
Do we need it?
</blockquote></details>

This just helps the case where we weren't able to reduce the size of the config even on the first attempt, so now if that happens we'd just treat that as a normal flushErr.


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7129 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cMTu6DGESSG7jG8FP:-N9cMTu6DGESSG7jG8FQ:b-gr8vr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/changefeed_test.go#L7129)):*
> ```Go
> 		})
> 
> 		t.Run(`fail permanently if individual messages are rejected by the server for being too large`, func(t *testing.T) {
> ```

If you're taking the markFatal() stuff out of this PR, I'd suggest taking this test out too or at least removing "permanently" from the description.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 614 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9ggPU01iPoLg2xO9rA:-N9ggPU01iPoLg2xO9rB:b-9kj71z) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L614)):*
> ```Go
> 	newConfig := *c
> 	newConfig.Producer.Flush.Messages = flooredHalve(c.Producer.Flush.Messages)
> 	newConfig.Producer.Flush.MaxMessages = flooredHalve(c.Producer.Flush.MaxMessages)
> ```

I think it's worth covering the case where Messages and MaxMessages started out at 0. This setting (a bit surprisingly) doesn't actually disable batching--if we call emitMessage() in rapid succession, Sarama will still batch (and potentially get the message size too large error from the server as a result). So it's possible to go "lower" than {0,0,0} for these three settings. It might work to do flooredHalve() for the other two and at the same time flooredHalveOrIfZeroThen1024() for MaxMessages (which also preserves the invariant that MaxMessages is always either 0 or >= Messages, which Sarama will error on otherwise).


<!-- Sent from Reviewable.io -->

samiskin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @livlobo, @miretskiy, and @samiskin)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cNY6KEAr0IrQfMI-K):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think pr description needs to be updated since retries are now batching.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XIYNP3AFKaYp9zAAx:-N9m5EtJ7OT2a4Fg4Ffo:b-tdsj0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/changefeed_test.go#L7146)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

nonblocking: this makes me wish we had either a version of assert/require that returned an error instead of taking a testing.T, or a version of SucceedsSoon that passed in a special Fataler so that you could just use require here.
</blockquote></details>

Yeah 😕
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7129 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cMTu6DGESSG7jG8FP:-N9m5A4i2BDoMWEtht9l:b-d9bgap) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/changefeed_test.go#L7129)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

If you're taking the markFatal() stuff out of this PR, I'd suggest taking this test out too or at least removing "permanently" from the description.
</blockquote></details>

I'm gonna leave the "make errors permanent" for a followup PR, for now I changed the description 👍
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 492 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9X4mVe33pq6GUWXrok:-N9m8_MW7LgOJ-5aRfgu:bs5vgzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L492)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This needs to be louder.  log.Error; plus add metrics.

Also, if the current batch size is 1 -- we definitely need to spit out permanent error.
</blockquote></details>

I'm gonna leave the "make errors permanent" for a followup PR
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 499 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XJP5MDnB9RNBUMG7s:-N9m7d6c7dNuIFQI4kP1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L499)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Unless you're doing this elsewhere I think we should check that ackMsg is a real message (has a non-nil Key and Value). We can also get errors for sentinel messages on this channel and those should not be retried.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XMvZAARAqvUXz1H7g:-N9m7fyH0CcAwPogoOSc:b-x670i3) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L557)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Does this end up swallowing the error or does it get surfaced somewhere?
</blockquote></details>

This gets surfaced as a clientErr that will then be returned on the next Flush
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cO-iH-kgI4a5K1Eo7:-N9m5JlR9qhy048_eu_M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L107)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

let's add a comment why we need this mutex.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 451 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cahyh-QxPHHlLq0JV:-N9m65YU3q9Cy6rquQ2e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L451)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

you might not need this variable if you initialize retryBuf in startInternalRetry; then check for retyrBuf!=nil
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 457 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cZf90586ptXeIowU9:-N9m66J89gIYx4Yhw7Zm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L457)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

as mentioned below, I'm concerned that early termination of this function may
result in lock being held, and caller unable to even notice the error (because it needs to acquire lock).
I think the following is safer:
```
type maybeLocker struct {
  wrapped sync.Locker
  locked bool
}
func (l *maybeLocker) Lock() {
   l.wrapped.Lock()
   l.locked = true
}
func (l *maybeLocker) Unlock() {
   if l.locked {
     l.wrapped.Unlock()
     l.locked = false
   }
}

...
clientMu := &maybeLocker{s.clientMu}
defer clientMu.Unlock()

startInternalRetry := func() { 
    clientMu.Lock()
}

```

Or.. if you don't like maybeLock helper, you can do that with functions...

</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 459 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cQLQA7Er9Pq2zd_-M:-N9m66qm6uKhTjUi7h62:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L459)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

lets 
```
s.clientMu.AssertHeld
```
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 464 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cZ1ND9D6hr42LFbBQ:-N9m67Oc46AqXeyNNTcR:b-j3cch6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L464)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should we propagate error here? We're "throwing" the client away  and going to replace...
I wonder if this should be silent; or perhaps log.error
</blockquote></details>

Made it log.Error for every close on an old client/producer
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 477 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cQU1Y4iTLpDeQwSNX:-N9m6DHlBFj5uo615E3r:b-sj5obh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L477)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

is this the only place .Config() is used? If so, consider just returning config form handleBuffeedRetries?
you could event pass current config (ptr) to handleBufferedRetries so that it mutates config directly,
w/out the need to create new config object.
</blockquote></details>

I use `.Config()` in the tests as well.  I'd rather not mutate config directly just cause this won't be called too often and I feel safer to future changes with things being more immutable.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 480 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cZBBLDDoVpXgCKJS4:-N9m6ZOr8K8Tq5fe5Sf9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L480)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I would prefer if this function (retryBufferedMessages) was removed, and 
this logic was done at the call site:
```
  if isBufferingRetries && s.mu.inflight == 0 {
     		validClient, validProducer, err := s.handleBufferedRetries(retryBuf, retryErr)
		if err != nil {
			s.clientMu.clientErr = err
			return
		}
                if err := s.updateClient(validClient, validProducer); err != nil {
                     ....
                }
  }
```

</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cR1TaDqNUmhBbb-pu:-N9m6cVbAiSWoZEeoFh8:b-jkkhwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L510)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm concerned that startInternalRetry grabs a lock, and until inflight == 0, we don't release the lock.
Yet, it's possible that some other error may occur, causing early termination of this go routine...

</blockquote></details>

Added the deferred unlock
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 564 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cUWqR8iWWa4hOZisp:-N9m7eWeAOiezJk97Up9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L564)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'd rather have reduceBatchingConfig return a bool indicating that it couldn't be reduced.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 585 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cUkrC-ZW7dvRpkR3C:-N9m8EZb7usseG_9qx6J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L585)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

definitely make it loud.  log error at least; 
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 606 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9cU53lEJFmZFU9Wrax:-N9m7uiZBEJh0UbR_5J5:b-2clp4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L606)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

does `return (num+1)>>1` work?
</blockquote></details>

It would but I'd rather do something with one less piece of mental processing needed.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 614 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9ggPU01iPoLg2xO9rA:-N9m821t4gphUFhgFBlr:b9lo8ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L614)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I think it's worth covering the case where Messages and MaxMessages started out at 0. This setting (a bit surprisingly) doesn't actually disable batching--if we call emitMessage() in rapid succession, Sarama will still batch (and potentially get the message size too large error from the server as a result). So it's possible to go "lower" than {0,0,0} for these three settings. It might work to do flooredHalve() for the other two and at the same time flooredHalveOrIfZeroThen1024() for MaxMessages (which also preserves the invariant that MaxMessages is always either 0 or >= Messages, which Sarama will error on otherwise).
</blockquote></details>

Ah good point, thought it was the case that at least one of them had to be 1, but turns out that's including Frequency which could be 1.  Made it set MaxMessages to 1000 if it was 0 since we default MaxMessages to 1000 earlier.


<!-- Sent from Reviewable.io -->

shermanCRL commented:
Further testing note, let’s exercise (try to break) the ordering guarantees. @miretskiy describes this as freezing.

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 492 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9X4mVe33pq6GUWXrok:-N9mS__e4M1IXF3ahZ4Z:becvbxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/sink_kafka.go#L492)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

I'm gonna leave the "make errors permanent" for a followup PR
</blockquote></details>

Can you add a todo somewhere?
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 555 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9c_4QWETtaed2aFJe8:-N9mYwqe1sb7dT5F7S-q:bcn15rm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L555)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

This just helps the case where we weren't able to reduce the size of the config even on the first attempt, so now if that happens we'd just treat that as a normal flushErr.
</blockquote></details>

I understand; but if you look at how scatter shot this retry error is.. i just dont' know if it's worth it...
would be nicer if error handling was localized instead.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 614 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9ggPU01iPoLg2xO9rA:-N9mZK3a7b_tZ_qnIjmf:b-dzuy5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L614)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

Ah good point, thought it was the case that at least one of them had to be 1, but turns out that's including Frequency which could be 1.  Made it set MaxMessages to 1000 if it was 0 since we default MaxMessages to 1000 earlier.
</blockquote></details>

Consider that 1000 is rather arbitrary; I'd rather do "less" but correctly.  Perhaps just focus on messges/maxmessages.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 125 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9mRgrN-1EY9whJO4s3:-N9mRgrN-1EY9whJO4s4:b-nt9lfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede0319378677e697ef8d732d1320fc272d0a32b/pkg/ccl/changefeedccl/sink_kafka.go#L125)):*
> ```Go
> 
> 	// Synchronized between the client and worker goroutines, where message emits
> 	// require access to the lock however if an internal retry process is
> ```

new sentence before however.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 482 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9mRyoO2d1XsHggroaJ:-N9mRyoO2d1XsHggroaK:bydn141) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede0319378677e697ef8d732d1320fc272d0a32b/pkg/ccl/changefeedccl/sink_kafka.go#L482)):*
> ```Go
> 	startInternalRetry := func(err error) {
> 		s.mu.AssertHeld()
> 		log.Infof(s.ctx, "kafka sink with flush config (%+v) beginning internal retry due to error: %s", s.kafkaCfg.Producer.Flush, err.Error())
> ```

let's do some wrap-o-magic
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 505 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9mST-dAM-Ivm15MGFi:-N9mST-dAM-Ivm15MGFj:b-jzqm70) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede0319378677e697ef8d732d1320fc272d0a32b/pkg/ccl/changefeedccl/sink_kafka.go#L505)):*
> ```Go
> 		// Log the active config every so often as it may be different from the
> 		// configured value if internal retries had ever been applied.
> 		if shouldLogConfig.ShouldLog() {
> ```

should we only do this when we change config?
and drop shouldLogConfig ?


<!-- Sent from Reviewable.io -->

samiskin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 555 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9c_4QWETtaed2aFJe8:-NA4t69tBQqBHWSS3X2F:blkjzxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L555)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I understand; but if you look at how scatter shot this retry error is.. i just dont' know if it's worth it...
would be nicer if error handling was localized instead.
</blockquote></details>

I feel like the alternative is a bit messier, where now an internally retryable error can be surfaced from multiple places and we're having to run an extra check on at least the first internally-retryable message, and on _every_ erroring message if we can't reduce the batching.

I like that this way its all a single flow of internally-retryable error detected -> attempt retry -> if it fails, surface error.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 614 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9ggPU01iPoLg2xO9rA:-NA4vEUc3Agh6XhiQwuY:b-jncga) ([raw file](https://github.com/cockroachdb/cockroach/blob/e38936b4ac60a0586b1343dfe31c3010f3e8283e/pkg/ccl/changefeedccl/sink_kafka.go#L614)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Consider that 1000 is rather arbitrary; I'd rather do "less" but correctly.  Perhaps just focus on messges/maxmessages.
</blockquote></details>

I reduced it to 250 just so that it's a smaller arbitrary number and removed the Bytes reduction to keep it simpler.  
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 125 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9mRgrN-1EY9whJO4s3:-NA4rQe9DUg3sGGSJjrY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede0319378677e697ef8d732d1320fc272d0a32b/pkg/ccl/changefeedccl/sink_kafka.go#L125)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

new sentence before however.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 482 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9mRyoO2d1XsHggroaJ:-NA4rXoY71U-oUO3gjay:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede0319378677e697ef8d732d1320fc272d0a32b/pkg/ccl/changefeedccl/sink_kafka.go#L482)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

let's do some wrap-o-magic
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 505 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9mST-dAM-Ivm15MGFi:-NA4rvenBifi5-ZI09E3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ede0319378677e697ef8d732d1320fc272d0a32b/pkg/ccl/changefeedccl/sink_kafka.go#L505)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should we only do this when we change config?
and drop shouldLogConfig ?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 1 of 3 files at r4, 1 of 3 files at r5, 1 of 2 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @samiskin)

___
*[`pkg/ccl/changefeedccl/metrics.go` line 423 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-NA50zrL4MqNmD3riZIr:-NA50zrL4MqNmD3riZIs:b-8fqzza) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ecc18278be17e4193ae167ebe6c86def93cda5f/pkg/ccl/changefeedccl/metrics.go#L423)):*
> ```Go
> 		Unit:        metric.Unit_BYTES,
> 	}
> 	metaBatchReductionCount := metric.Metadata{
> ```

do you think that perhaps we should keep track of "messages per batch" metric instead of (or maybe in addition) to this metric?
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 619 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-NA51NMR5GUHsEc9ZVxI:-NA51NMR5GUHsEc9ZVxJ:b-g22172) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ecc18278be17e4193ae167ebe6c86def93cda5f/pkg/ccl/changefeedccl/sink_kafka.go#L619)):*
> ```Go
> 		s.metrics.recordBatchReduction()
> 
> 		log.Infof(s.ctx, "kafka sink retrying %d messages with reduced flush config: (%+v)", len(msgs), newConfig.Producer.Flush)
> ```

nit: a bit of  wrapping please (if needed... perhaps just move args to the next line)
___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 680 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-NA51_y14rzOYebT9UK0:-NA51_y14rzOYebT9UK1:b-x3utth) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ecc18278be17e4193ae167ebe6c86def93cda5f/pkg/ccl/changefeedccl/sink_kafka.go#L680)):*
> ```Go
> 	}
> 
> 	wasReduced := newConfig.Producer.Flush == c.Producer.Flush
> ```

did you mean Flush != ?


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86138)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @samiskin)

___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86138#-N9XIYNP3AFKaYp9zAAx:-NAAI5cFAeNIvq9mr62o:b81y4zt) ([raw file](https://github.com/cockroachdb/cockroach/blob/eebc9d46a07958842e7657e82b75f698e8fd1ec9/pkg/ccl/changefeedccl/changefeed_test.go#L7146)):*
<details><summary><i>Previously, samiskin (Shiranka Miskin) wrote…</i></summary><blockquote>

Yeah 😕
</blockquote></details>

(Going to write this btw)


<!-- Sent from Reviewable.io -->

samiskin commented:
bors r+

samiskin commented:
blathers backport 22.1
